zarr/storage.py

```diff
@@ -334,6 +337,19 @@ def _init_array_metadata(store, shape, chunks=None, dtype=None, compressor='defa
     if filters:
         filters_config = [f.get_config() for f in filters]
     else:
+        filters_config = []
+
+    # deal with object encoding
+    if dtype == object:
+        if object_codec is None:
+            raise ValueError('an object_codec is required for object arrays')
+        else:
+            filters_config.insert(0, object_codec.get_config())
+    elif object_codec is not None:
+        raise ValueError('an object_codec is only needed for object arrays')
```

Comment:
Maybe this should be softened to a warning? While it certainly isn't sensible for a user to set `object_codec` when there is no `object`, there isn't any problem caused by adding it. We just ignore it in this case anyways.

We could also explicitly set `object_codec` to `None` if we make this a warning should it accidentally be used later.

Reply:
Yeah, I did wonder about that. In the end I figured 6 of one, half dozen of the other. Also I was being slightly lazy, I hate testing warnings, have had awful PY2/PY3 compatibility issues.

Reply:
So as a friendly tip about testing warnings. Normally I having the `warnings` module convert them to errors and then test them in exact same way. The [last half of this SO comment]( https://stackoverflow.com/a/15934081 ) is useful for showing this. When we switch to `pytest`, they have a [nicer way of doing this]( https://docs.pytest.org/en/latest/warnings.html#warns ).

Reply:
Thanks, I will soften to warning. I have some other tests that test warnings by converting to errors, but I could never get the ``catch_warnings`` context manager to work in PY2 so there is a bunch of ugly calls to ``warnings.resetwarnings()``. Maybe this is time to start using pytest if that solutions works in PY2 and PY3.